crates/typst-cli/src/terminal.rs

```diff
@@ -0,0 +1,140 @@
+use std::io::{self, IsTerminal, Write};
+use std::sync::atomic::{AtomicBool, Ordering};
+
+use codespan_reporting::term::termcolor;
+use ecow::eco_format;
+use once_cell::sync::Lazy;
+use termcolor::{ColorChoice, WriteColor};
+use typst::diag::StrResult;
+
+use crate::ARGS;
+
+/// Returns a handle to the optionally colored terminal output.
+pub fn out() -> TermOut {
+    static OUTPUT: Lazy<TermOutInner> = Lazy::new(TermOutInner::new);
+    TermOut { inner: &OUTPUT }
+}
+
+/// The stuff that has to be shared between instances of [`TermOut`].
+struct TermOutInner {
+    active: AtomicBool,
+    stream: termcolor::StandardStream,
+    in_alternate_screen: AtomicBool,
+}
+
+impl TermOutInner {
+    fn new() -> Self {
+        let color_choice = match ARGS.color {
+            clap::ColorChoice::Auto if std::io::stderr().is_terminal() => {
+                ColorChoice::Auto
+            }
+            clap::ColorChoice::Always => ColorChoice::Always,
+            _ => ColorChoice::Never,
+        };
+
+        let stream = termcolor::StandardStream::stderr(color_choice);
+        TermOutInner {
+            active: AtomicBool::new(true),
+            stream,
+            in_alternate_screen: AtomicBool::new(false),
+        }
+    }
+}
+
+/// A utility that allows users to write colored terminal output.
+/// If colors are not supported by the terminal, they are disabled.
+/// This type also allows for deletion of previously written lines.
+#[derive(Clone)]
+pub struct TermOut {
+    inner: &'static TermOutInner,
+}
+
+impl TermOut {
```

Comment:
why do some of `TermOut`'s methods take `&mut self`?

also, why do we have `TermOut` and `TermOutInner` instead of just letting `out()` return an `&'static TermOut`?

edit: probably because Write requires `&mut self`? perhaps `out()` could return a locked TermOut like `std::io::stdout()` does?

Reply:
I agree with the suggestion to return a locked handle. Locking the mutex in every write call is sub-optimal.

Reply:
> why do some of `TermOut`'s methods take `&mut self`?

I thought it would be clearer which methods mutate things like that. It doesn't change anything in actual usage.

> also, why do we have `TermOut` and `TermOutInner` instead of just letting `out()` return an `&'static TermOut`?
> 
> edit: probably because Write requires `&mut self`? perhaps `out()` could return a locked TermOut like `std::io::stdout()` does?

Yes, it's necessary because of `Write's` and `WriteColor`s `&mut self` parameter.
`std::io::stdout()` does not actually return a locked handle. One needs to call `.lock()` on it.

I'm not sure always locking is benefitial here:
- The write methods are only rarely called, so the performance benefit would be neglegible.
- The `.is_active()` method (among others) does not require locking. It is very regularly called. It's not a huge performance hit to lock here, but not very nice either.
- Locking on every `terminal::out()` call might cause a deadlock if insufficient care is taken when writing the code.

The alternative would be an extra `.lock()` method like `std::io::stdout()` and `termcolor::StandardStream` offer, but I don't think there is enough of a case to warrant the extra code required.

I therefore think it is preferable to implicitly lock in every write call, even though it does add a small performance overhead.

Reply:
Okay